### ISSUE TITLE
Scripting: TileLayerEdit.setTile() is affected by current selection

### ISSUE THREAD

**Issue by eishiya** [mapeditor/tiled#3482]
If the current map has any tiles selected, `TileLayerEdit.setTile()` will only work within the selection, and will not change any tiles outside the selection. This means that it is necessary to clear the user's selection when it isn't relevant to the script's action. Scripts should not have to do that.

I expect `setTile()` to set the tile, nothing more, nothing less. Its documentation does not state anything else. It also ignores layer lock status, as it should! It should be up to scripts to respect restrictions like locks and selections.

That said, a version of `setTile()` that *does* automatically respect selections (and layer locks!) could be convenient too, as a separate function that returns `true` when it set the tile and `false` when it did not.

**Comment by firstmiddlelast** [comment]
Agreed. At least, setTile() should return an indication whether it actually changed something or not.

**Comment by bjorn** [comment]
To be a bit more precise, it's not `TileLayerEdit.setTile()` that only works within the selection, it's `TileLayerEdit.apply()` that does this, because it internally uses the `PaintTileLayer` undo command, which in turn uses `TilePainter::setCells`, which takes into account the selection (as well as map size). It's happening in that helper class, because taking into account the selection and map size is not something I wanted to code into all the various painting tools, but should work the same way for all of them.

I guess the easiest approach here would be to add an option to `TilePainter` to ignore the selection, and to mirror this option on the `PaintTileLayer` command and the `TileLayerEdit` object (similar to the `mergeable` property, we could have `ignoreSelection`). Actually providing feedback about whether a change was made would be harder, but if a script really needs to know that it could just check whether the changed region intersects with the selection on its own (a `TileLayerEdit.changedRegion` property could be added to make this more convenient).

Since `TileLayerEdit` can only modify a single layer at a time, and tile editing tools can provide a `TileMap` to show a preview, another common way to trigger changes from scripted tools is through `TileMap.merge` (though it only works on open assets). This one has a parameter `bool canJoin`... we could add a `bool ignoreSelection` there, but I'm not really happy with multiple Boolean parameters, so maybe an overload could be added instead that takes some flags.

**Comment by bjorn** [comment]
> That said, a version of `setTile()` that _does_ automatically respect selections (and layer locks!) could be convenient too, as a separate function that returns `true` when it set the tile and `false` when it did not.

The `TileMap.merge` function does take layer locks into account, but indeed `TileLayerEdit.apply` currently does not. I guess the latter could just be a no-op when its target layer is locked, unless a new `ignoreLock` option is set to `true`.

Or maybe we could combine `ignoreSelection` and `ignoreLock` options into a single `force` option?
